cpp/src/arrow/array/builder_binary.h

```diff
@@ -51,10 +51,12 @@ class BaseBinaryBuilder : public ArrayBuilder {
   using offset_type = typename TypeClass::offset_type;
 
   explicit BaseBinaryBuilder(MemoryPool* pool = default_memory_pool())
-      : ArrayBuilder(pool), offsets_builder_(pool), value_data_builder_(pool) {}
+      : ArrayBuilder(pool), offsets_builder_(pool),
+        value_data_builder_(pool), type_(binary()) {}
 
   BaseBinaryBuilder(const std::shared_ptr<DataType>& type, MemoryPool* pool)
-      : BaseBinaryBuilder(pool) {}
+      : ArrayBuilder(pool), offsets_builder_(pool),
+        value_data_builder_(pool), type_(type) {}
```

Comment:
Please make this constructor `protected`; this will make it clearer that BaseBinaryBuilder is intended for use by subclasses only

Reply:
Done